src/acq.c

```diff
@@ -34,969 +39,1914 @@
-	acq_g.chan[channel].sample32 = sample32;
 
-	return BL_ERROR_NONE;
 }
 
-/**
- * Set the hardware up for an acquisition.
- *
- * \param[in]  frequency   Sample frequency in Hz.
- * \param[in]  oversample  Number of bits to oversample by.
- * \param[in]  src_mask    Mask of sources to enable.
- * \return \ref BL_ERROR_NONE on success, or appropriate error otherwise.
- */
-static enum bl_error bl_acq_setup(
+/* Exported function, documented in acq.h */
+enum bl_error bl_acq_start(
 		uint16_t frequency,
 		uint32_t oversample,
 		uint16_t src_mask)
 {
-	enum bl_error error;
+	if (src_mask == 0x00) {
+		return BL_ERROR_BAD_SOURCE_MASK;
+	}
+
+	if (src_mask >= (1U << BL_ACQ__SRC_COUNT)) {
+		return BL_ERROR_BAD_SOURCE_MASK;
+	}
+
+	if (bl_acq_is_active) {
+		return BL_ERROR_ACTIVE_ACQUISITION;
+	}
 
 	for (unsigned i = 0; i < BL_ACQ__SRC_COUNT; i++) {
-		if (acq_g.chan[i].gain == 0) {
-			acq_g.chan[i].gain = 1;
+		bl_acq_channel_t *channel = &bl_acq_channel[i];
+		bl_acq_channel_config_t *config = &channel->config;
+
+		/* Finalize Config. */
+		config->enable = ((src_mask & (1U << i)) != 0);
+		if (!config->enable) {
+			continue;
+		}
+
+		config->frequency     = frequency;
+		config->sw_oversample = oversample;
+		config->oversample    = 0;
+
+		config->frequency_trigger = (config->frequency *
+				config->sw_oversample);
+		config->frequency_sample = (config->frequency_trigger <<
+				config->oversample);
+
+		bool opamp_needed = (config->opamp_gain > 1) ||
+			(config->opamp_offset != OPAMP_OFFSET_NONE);
+		if (opamp_needed && (channel->opamp == NULL)) {
+			return BL_ERROR_HARDWARE_CONFLICT;
+		}
+
+		config->opamp_used = opamp_needed || (channel->adc == NULL);
+
+		/* Initialize message queue. */
+		channel->msg = bl_mq_acquire(i);
+		if (channel->msg != NULL) {
+			channel->msg->type = channel->config.sample32 ?
+				BL_MSG_SAMPLE_DATA32 : BL_MSG_SAMPLE_DATA16;
+			channel->msg->sample_data.channel  = i;
+			channel->msg->sample_data.count    = 0;
+			channel->msg->sample_data.reserved = 0x00;
 		}
 	}
 
-	/* Calculate prescale required with current oversample settings */
-	uint32_t period, prescale;
-	bl_frequency_to_acq_params(frequency, oversample, &prescale, &period);
+	/* Configure timer(s). */
+	for (unsigned i = 0; i < BL_ACQ__SRC_COUNT; i++) {
+		bl_acq_channel_t *channel = &bl_acq_channel[i];
+		bl_acq_timer_t *timer = bl_acq_channel_get_timer(channel);
+
+		if (timer == NULL) {
+			continue;
+		}
+
+		/* Ensure same frequency for all users of the timer. */
+		for (unsigned j = (i + 1); j < BL_ACQ__SRC_COUNT; j++) {
+			bl_acq_channel_t *cb = &bl_acq_channel[j];
+			if (bl_acq_channel_get_timer(cb) != timer) {
+				continue;
+			}
+
+			if (channel->config.frequency_trigger !=
+					cb->config.frequency_trigger) {
+				return BL_ERROR_HARDWARE_CONFLICT;
+			}
+		}
 
-	error = bl_acq_setup_timer(period, prescale);
-	if (error != BL_ERROR_NONE) {
-		return error;
+		enum bl_error status = bl_acq_timer_configure(timer,
```

Comment:
shall we only declare status once and use it all through the function?

Reply:
I personally prefer to keep the declarations as close to use as possible and avoid re-using the same declaration, but it'd be fine to declare once as well.